### PR TITLE
Advanced TM validation Part 2

### PR DESCRIPTION
**Pull request by fatadel**

### CODE REVIEW

File: packages/core/shared.js [hunk 1]
```diff
@@ -1135,3 +1140,62 @@ async function checkLinkedAffordances(td) {
...
     return results;
 }
+
+/**
+ * Given a TD check that its structure corresponds to the one imposed by the linked TM.
+ *
+ * @param {object} td - TD to check
+ */
+async function checkLinkedStructure(td) {
+    // TODO: What's the assertion name?
```
Comment (egekorkan) [inline packages/core/shared.js:1150]: thing-model-td-generation-processor-imports should be used

**Review by egekorkan — CHANGES_REQUESTED** [review]

**Comment by egekorkan** [comment]
The TD examples that use TMs should be updated accordingly

File: packages/core/shared.js [hunk 2]
```diff
@@ -965,72 +967,49 @@ function checkTmOptionalPointer(td){
...
-            {
-                ID: ASSERTION_OPTIONAL,
-                Status: 'not-impl',
-                Comment: 'td does not link to tm or links to more than one tm'
-            }
-        ];
+        return {
+            success: false,
+            status: 'not-impl',
+            comment: 'td does not link to tm or links to more than one tm'
```
[the remark this answers is not in the document]
Comment (fatadel) [reply packages/core/shared.js]: ✅

File: packages/core/shared.js (hunk 1 above)
Comment (fatadel) [reply packages/core/shared.js:1154]: ✅

**Comment by egekorkan** [comment]
A login is needed to see the netlify error and it is 
Screenshot 2023-01-05 at 13 38 08

This is due to one of our dependency's dependency so nothing too much we can do but good to know.

**Comment by egekorkan** [comment]
Locally it works. There is one major question that popped to my mind though: 

If a key imposed by a TM is missing, an error is thrown. However, if the value of that key is different, no error is thrown. For example, if a property affordance has type string in a TM, it should not be possible to say that it is a number in the TD. In a specific example of https://github.com/thingweb/thingweb-playground/blob/advanced-tm-validation/packages/core/examples/tds/valid/usingTM.json , the status property cannot be changed into a number. 

I will however still merge it and open that as a separate issue.

**Comment by fatadel** [comment]
> Locally it works. There is one major question that popped to my mind though:
> 
> If a key imposed by a TM is missing, an error is thrown. However, if the value of that key is different, no error is thrown. For example, if a property affordance has type string in a TM, it should not be possible to say that it is a number in the TD. In a specific example of https://github.com/thingweb/thingweb-playground/blob/advanced-tm-validation/packages/core/examples/tds/valid/usingTM.json , the status property cannot be changed into a number.
> 
> I will however still merge it and open that as a separate issue.

Correct, it was intended to do this work in 2 steps - (1) check the keys only, (2) check keys and values. I just did not make it in time before the break to implement the second part. Starting right away with that.

**Pull request merged**